src/Microsoft.TemplateEngine.Cli/Commands/create/InstantiateCommand.TabCompletion.cs

```diff
@@ -110,5 +118,55 @@ protected internal override IEnumerable<CompletionItem> GetCompletions(Completio
                 yield return completion;
             }
         }
+
+        private static IEnumerable<CliTemplateInfo> GetAllowedTemplates(TemplateConstraintManager constraintManager, TemplateGroup templateGroup)
+        {
+            if (templateGroup.Templates.SelectMany(t => t.Constraints).Any())
+            {
+                var timeoutTask = Task.Delay(ConstraintEvaluationTimeout);
+                var constraintEvaluationTask = templateGroup.GetAllowedTemplatesAsync(constraintManager, default);
+                Task.Run(async () => await Task.WhenAny(constraintEvaluationTask, timeoutTask).ConfigureAwait(false)).GetAwaiter().GetResult();
```

Comment:
Since this is for the CLI (and therefore not stuck on net472-available APIs), instead of the 'racing tasks' model you may be able to use WaitAsync(TimeSpan), WaitAsync(CancellationToken), or even using a CancellationTokenSource with a timespan + the cancellationtoken from that source instead of the `default` cancellationToken. Same with `GetallowedTemplateGroups` below.